File: unid-node-wallet-sdk/src/core/payload.ts
import {
    DIDCreatePayload,
    DIDCreateRequest,
    DIDReplaceAction,
    DIDReplaceDeltaObject,
    DIDReplacePayload,
    DIDReplaceSuffixObject,
} from "./interfaces/did-operation";
import { Encoder } from "../utils/encoder";
import { Multihash } from "../utils/multihash";

/**
 * @param params 
 * @returns
 */
export const didCreatePayload = (params: DIDCreateRequest): DIDCreatePayload => {
    const document: DIDReplacePayload = {
        public_keys: params.publicKeys,
        service_endpoints: [],
    }
    const patch: DIDReplaceAction = {
        action: 'replace',
        document: document
    }

    const delta: DIDReplaceDeltaObject = {
        patches: [ patch ],
        update_commitment: Multihash.canonicalizeThenDoubleHashThenEncode(params.commitmentKeys.update),
    }
    const deltaBuffer = Buffer.from(JSON.stringify(delta))
    const deltaHash   = Encoder.encode(Multihash.hash(deltaBuffer))
    
    const suffixData: DIDReplaceSuffixObject = {
        delta_hash: deltaHash,
        recovery_commitment: Multihash.canonicalizeThenDoubleHashThenEncode(params.commitmentKeys.recovery)
    }
    const deltaEncodedString = Encoder.encode(deltaBuffer);
    const suffixDataEncodedString = Encoder.encode(JSON.stringify(suffixData));

    const payload: DIDCreatePayload = {
        type: 'create',
        delta: deltaEncodedString,
        suffix_data: suffixDataEncodedString,
    }

    return payload
}

File: src/core/payload.rs
use super::super::utils::{encoder::*, multihasher::*};
use super::interfaces::did_operation::*;
pub struct Payload {}

impl Payload {
  pub fn did_create_payload(params: DIDCreateRequest) -> DIDCreatePayload {
    let document: DIDReplacePayload = DIDReplacePayload {
      public_keys: params.public_keys,
      service_endpoints: Vec::new(),
    };

    let patch: DIDReplaceAction = DIDReplaceAction {
      action: "replace".to_string(),
      document,
    };

    let delta: DIDReplaceDeltaObject = DIDReplaceDeltaObject {
      patches: vec![patch],
      update_commitment: Multihasher::canonicalize_then_double_hash_then_encode(
        serde_json::to_value(&params.commitment_keys.update).unwrap(),
      ),
    };

    let delta_serde: serde_json::Value = serde_json::to_value(&delta).unwrap();
    let delta_str: &str = &delta_serde.to_string();
    let delta_u8: &[u8] = delta_str.as_bytes();
    let delta_buf: Vec<u8> = delta_u8.to_vec();
    let delta_hash: String = Encoder::encode(Multihasher::hash(delta_buf.to_owned()));
    let suffix_data: DIDReplaceSuffixObject = DIDReplaceSuffixObject {
      delta_hash,
      recovery_commitment: Multihasher::canonicalize_then_double_hash_then_encode(
        serde_json::to_value(&params.commitment_keys.update).unwrap(),
      ),
    };
    let delta_encoded_string: String = Encoder::encode(delta_buf);
    let suffix_serde: serde_json::Value = serde_json::to_value(&suffix_data).unwrap();
    let suffix_str: &str = &suffix_serde.to_string();
    let suffix_u8: &[u8] = suffix_str.as_bytes();
    let suffix_buf: Vec<u8> = suffix_u8.to_vec();
    let suffix_data_encoded_string: String = Encoder::encode(suffix_buf);

    let payload: DIDCreatePayload = DIDCreatePayload {
      type_field: "create".to_string(),
      delta: delta_encoded_string,
      suffix_data: suffix_data_encoded_string,
    };

    payload
  }
}
